### PR TITLE
Add F5 BIG IP Mixin

### PR DESCRIPTION
**Pull request by JonathanWamsley**
Note, this dashboard will not support a k8s plugin.

Dashboards include:
-  F5 cluster overview
-  F5 node overview
-  F5 pool overview
-  F5 virtual server overview
-  F5 logs overview

Alerts include:
- F5BigIPLowNodeAvailabilityStatus
- F5BigIPServerSideConnectionLimit
- F5BigIPHighRequestRate
- F5BigIPHighConnectionQueueDepth

Cluster overview
bigip-cluster-overview

Node overview

bigip-node-overview

Pool overview
bigip-pool-overview-1

bigip-pool-overview-2

Virtual server overview

bigip-virtual-server-overview-1
bigip-virtual-server-overview-2

Log overview
bigip-logs-overview

### CODE REVIEW

File: f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet
Comment (dpaasman00) [inline f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet:138]: think this unit should be changed unless you want to make this a rate panel, which i think you easily could, your call

File: f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet
Comment (dpaasman00) [inline f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet:156]: believe this panel will need to include a mean/min/max legend table since there are 3+ metrics/queries

File: f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet
Comment (dpaasman00) [inline f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet:460]:
```suggestion
  title: 'Packets / $__interval',
```

File: f5-bigip-mixin/dashboards/f5-node-overview.libsonnet
Comment (dpaasman00) [inline f5-bigip-mixin/dashboards/f5-node-overview.libsonnet:539]: does it make sense to standardize how this 3rd selector is queried across the node, pool, and vs dashboards? Like the pool and vs selector queries are similar on their respective dashboards where as this one is done differently. Just curious if it makes sense to match them all up?

**Review by dpaasman00 — COMMENTED** [review]
First pass pretty solid, just some nits and questions

File: f5-bigip-mixin/README.md
Comment (JonathanWamsley) [reply f5-bigip-mixin/README.md]: updated

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:45]: updated

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet]: good catch, updated

File: f5-bigip-mixin/dashboards/f5-node-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-node-overview.libsonnet]: updated

File: f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet]: updated to none instead of requests.

File: f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet:156]: added a panel table

File: f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-virtual-server-overview.libsonnet]: updated

File: f5-bigip-mixin/dashboards/f5-node-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-node-overview.libsonnet]: Ahh I see, yeah, there is no reason for the or statement in to include the job or the instance here. Updated

File: f5-bigip-mixin/README.md
Comment (StefanKurek) [inline f5-bigip-mixin/README.md:5]: Just thinking, when abbreviating how we refer to "F5 Big IP" is it more appropriate to say "Big IP" rather than "F5"? Should these dashboards be "Big IP cluster overview", "Big IP node overview", etc. 

For example, for Apache Tomcat we wouldn't say "Apache overview" if we shortened it. We'd say "Tomcat overview" right?

File: f5-bigip-mixin/alerts/alerts.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/alerts/alerts.libsonnet:10]: is availability state only a 0 or 1? i thought nodes could have more than 2 states?

File: f5-bigip-mixin/alerts/alerts.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/alerts/alerts.libsonnet:50]: should this be critical? it seems like this could happen normally maybe?  like if there's not much traffic on a site. and then there is a surge because of the time of day...would this trigger?

File: f5-bigip-mixin/config.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/config.libsonnet:15]: is this even possible? how would the requests over the last 10 minutes / requests over the last 60 minutes (which includes that 10 min windows) ever be greater than 100%?

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:17]: general question, would it be better also include "job" across all aggregations like sum (for all dashboards)?

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:78]: just double checking that this state metric can only be 0 or 1 (i recall there being more than 2 states)

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:139]: same question about states

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:354]: increase means you have to change the title right? watch for this throughout

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (StefanKurek) [inline f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:588]: don't we usually capitalize the first letter of these labels (sentence case)?

**Review by StefanKurek — CHANGES_REQUESTED** [review]
Light review. I'm out of the mix and @Caleb-Hurshman or @BominRahmani can probably provide better feedback on formatting standards than I can at this point.

File: f5-bigip-mixin/README.md
Comment (JonathanWamsley) [reply f5-bigip-mixin/README.md:5]: Yeah, that makes sense. Updating F5 to Big IP or bigip when F5 is used alone.

File: f5-bigip-mixin/alerts/alerts.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/alerts/alerts.libsonnet:10]: Only 0 or 1.

File: f5-bigip-mixin/alerts/alerts.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/alerts/alerts.libsonnet:50]: yeah, if there is hardly any traffic and then a check-in starting time for a particular application, then the threshold would have a surge suddenly. I can switch this to warning. (This is with the offset or a rate thought process)

File: f5-bigip-mixin/config.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/config.libsonnet:15]: This is not possible with an increase function since the 60 minutes includes the last 10 minutes and therefore the numerator can never be higher than the denominator and be greater than 100 (assuming no negatives). There was some confusion going back and fourth between rate and increase here. One way to fix this while still using an increase is by using an offset `increase(bigip_pool_tot_requests[50m] offset 10m)` . I'll simulate some load tomorrow to get a better picture of this.

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:17]: yeah

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:78]: The [bigip exporter](https://github.com/ExpressenAB/bigip_exporter/blob/e0f009c578b6264ea6c2fb82071ceb02a1a8121b/collector/vs_collector.go#L450) looks for "available" and returns 1 if found. Everything else is a 0 in this case.

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:139]: No states label in this case

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:354]: updated

File: f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/f5-cluster-overview.libsonnet:588]: I think I started without the linter and applied the suggestion. Guess that is why we exclude the `template-job-rule` to get it to pass.
Screenshot 2023-09-19 at 4 43 20 PM

Updated

File: f5-bigip-mixin/README.md
Comment (Caleb-Hurshman) [inline f5-bigip-mixin/README.md:129]: Same thing here, new standard for time period.

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (Caleb-Hurshman) [inline f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet:340]: Because this panel displays >3 metrics, it should include min, mean, and max on the table. This also applies to the Connections panel and the Packets panel. I would recommend putting the legend to the right in this case for a more organized look.

File: f5-bigip-mixin/config.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/config.libsonnet:15]: testing this out more with some load, rate is the only way forward with this approach. updated to rate.

**Review by Caleb-Hurshman — CHANGES_REQUESTED** [review]
Looks good, just a few small tweaks to make as a result of new standards.

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet:340]: thanks, updated those charts to include a table to the right.

File: f5-bigip-mixin/README.md
Comment (JonathanWamsley) [reply f5-bigip-mixin/README.md]: updated

File: f5-bigip-mixin/config.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/config.libsonnet]: updated

**Review by Caleb-Hurshman — APPROVED** [review]
LGTM ✅

**Review by dpaasman00 — APPROVED** [review]
LGTM

**Review by StefanKurek — APPROVED** [review]

**Review by mshahzeb — CHANGES_REQUESTED** [review]
LGTM Overall! Just some suggestions

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (Dasomeone) [inline f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet:340]: This absolutely needs to be split into at least two panels for `inbound` and `outbound`, possibly even with four if you separate out ephemerals.

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (Dasomeone) [inline f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet:437]: This absolutely needs to be split into at least two panels for `inbound` and `outbound`, possibly even with four if you separate out ephemerals.

**Review by Dasomeone — CHANGES_REQUESTED** [review]

File: f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet:1]: I'll scale down all the panels in height. Let me know if this looks better. I'll update the screenshots afterwards if it looks acceptable.

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet:1]: Made it wide, and followed same pattern of splitting into 2 panels for the ephemeral metrics.

File: f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet]: updated to gauge

File: f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet:573]: fixed

File: f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet:732]: fixed

File: f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet:897]: fixed

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet:1135]: fixed

File: f5-bigip-mixin/dashboards/bigip-logs-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-logs-overview.libsonnet:1]: fixed

File: f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet]: split metrics inbound and outbound for all traffic panels.

File: f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet]: split metrics inbound and outbound for all packets panels.

File: f5-bigip-mixin/.lint
Comment (JonathanWamsley) [reply f5-bigip-mixin/.lint]: updated

File: f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet:1]: updated

File: f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet]: updated, lengthened with right aligned table legend

File: f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet:336]: updated, lengthened with right aligned table legend.

File: f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet]: updated

File: f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-pool-overview.libsonnet]: updated

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet:1]: updated

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet]: updated splitting ephemeral too

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet]: updated splitting ephemeral too

File: f5-bigip-mixin/dashboards/bigip-logs-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-logs-overview.libsonnet]: updated

File: f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet:1]: updated to table for those 3 dashboards. I played with a bunch of options to make it look good.

Comment (Dasomeone) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet:1]: Great change, I just think it needs to be a tad wider?
At least on smaller monitors, you might end up with something like this:
image

File: f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet
Comment (Dasomeone) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet]: Nice, thank you! Looks much better :D

File: f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet
Comment (Dasomeone) [reply f5-bigip-mixin/dashboards/bigip-virtual-server-overview.libsonnet]: Awesome, this looks great!

File: f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet
Comment (Dasomeone) [reply f5-bigip-mixin/dashboards/bigip-cluster-overview.libsonnet:1]: It still looks a bit large to me, with the text larger than the title. 
I had a play around, and the `text: size` options don't seem to have the largest impact unfortunately :(

It appears that what has an actual impact is panel height, as it scales the text appropriately then, however that's hard to then predict as we can't well just scale it to two nodes as in the sample app currently...
Example before/after
image

image


**Tl;dr: I think we have to live with it as-is. @mshahzeb what do you think?**

**Review by Dasomeone — APPROVED** [review]
Thanks for iterating on this Jon! Looks much better :D

Still would love for @mshahzeb and possibly @rgeyer to chime in on the label discussion re: `partition` but it's an approval from my end :)

File: f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet
Comment (JonathanWamsley) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet:1]: I updated the tables from 'center' alignment (good till 1300 pixels wide at 100% screen size resolution before collapse) to 'left' alignment (good till 1090 pixels at 100%).

Comment (Dasomeone) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet:1]: Yep, that works! Great way to fix it too :D

**Review by mshahzeb — APPROVED** [review]
LGTM! Thanks for the updates!

File: f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet
Comment (rgeyer) [reply f5-bigip-mixin/dashboards/bigip-node-overview.libsonnet:1]: My vote would be to add a selector which defaults to "all" and remove it from all of the legends. The queries would have to aggregate it away also if they don't already.

From what I can tell, partitions are basically just [ACLs in the BigIP config space](https://techdocs.f5.com/kb/en-us/products/big-ip_ltm/manuals/product/tmos-concepts-11-5-0/8.html).

So from an operational perspective I would very rarely care which partition has a given metric/trend, unless I were specifically looking for it. I.E. We know that the "Staging" partition is experiencing high latency or dropped packets, so lets filter to that.

**Review by rgeyer — APPROVED** [review]
Looks good! I weighed in on the question of having `partition` as part of the legend for all panels as well.

**Comment by JonathanWamsley** [comment]
> > Hey @ryan-grafana , F5 BIG-IP mixin is ready to be looked at! Thanks
> 
> Wrong Ryan, but.. I got it. ;)

my bad @rgeyer 😅, I think it is ready to merge when ready! Thank you 🙏

**Pull request merged**
